File: desktop/electron/hls-manifest.js
"use strict";

// Leitura mínima de manifestos HLS para comparar servidores antes de escolher um.
// Espelha android/app/src/main/java/com/obaflix/bridge/HlsManifest.kt — os dois
// extratores precisam pontuar as mesmas fontes da mesma forma.

function looksLikeManifest(text) {
  return String(text || "").trimStart().startsWith("#EXTM3U");
}

/**
 * Divide a lista de atributos de uma tag EXT-X preservando vírgulas dentro de
 * valores entre aspas, como em CODECS="avc1.4d401f,mp4a.40.2".
 */
function splitAttributes(raw) {
  const parts = [];
  let current = "";
  let quoted = false;
  for (const char of String(raw || "")) {
    if (char === '"') {
      quoted = !quoted;
      current += char;
    } else if (char === "," && !quoted) {
      parts.push(current);
      current = "";
    } else {
      current += char;
    }
  }
  if (current) parts.push(current);

  const values = new Map();
  for (const part of parts) {
    const eq = part.indexOf("=");
    if (eq <= 0) continue;
    const key = part.slice(0, eq).trim().toUpperCase();
    const value = part.slice(eq + 1).trim().replace(/^"|"$/g, "");
    if (key) values.set(key, value);
  }
  return values;
}

function qualityLabel(resolution, bandwidth) {
  const height = Number.parseInt(String(resolution || "").split("x")[1], 10);
  if (Number.isFinite(height) && height > 0) return `${height}p`;
  if (bandwidth > 0) return `${Math.round(bandwidth / 1000)} kbps`;
  return "auto";
}

function parse(text, baseUrl) {
  if (!looksLikeManifest(text)) return { isMaster: false, variants: [], audioTracks: [], subtitles: [] };

  const variants = [];
  const audio = new Set();
  const subtitles = new Map();

  for (const line of String(text).replace(/\r\n?/g, "\n").split("\n")) {
    const trimmed = line.trim();

    if (/^#EXT-X-STREAM-INF:/i.test(trimmed)) {
      const attrs = splitAttributes(trimmed.slice(trimmed.indexOf(":") + 1));
      const bandwidth =
        Number.parseInt(attrs.get("BANDWIDTH") || attrs.get("AVERAGE-BANDWIDTH") || "0", 10) || 0;
      const resolution = attrs.get("RESOLUTION") || null;
      variants.push({ bandwidth, resolution, label: qualityLabel(resolution, bandwidth) });
      continue;
    }

    if (/^#EXT-X-MEDIA:/i.test(trimmed)) {
      const attrs = splitAttributes(trimmed.slice(trimmed.indexOf(":") + 1));
      const name = attrs.get("NAME") || attrs.get("LANGUAGE") || null;
      const type = (attrs.get("TYPE") || "").toUpperCase();
      if (type === "AUDIO") {
        if (name) audio.add(name);
      } else if (type === "SUBTITLES" || type === "CLOSED-CAPTIONS") {
        const uri = attrs.get("URI");
        if (!uri) continue;
        try {
          const resolved = new URL(uri, baseUrl).toString();
          if (!subtitles.has(resolved)) {
            subtitles.set(resolved, { file: resolved, label: name || "Legenda", kind: "captions" });
          }
        } catch { /**/ }
      }
    }
  }

  const seenLabels = new Set();
  const distinct = variants
    .filter((variant) => {
      if (seenLabels.has(variant.label)) return false;
      seenLabels.add(variant.label);
      return true;
    })
    .sort((a, b) => b.bandwidth - a.bandwidth);

  return {
    isMaster: variants.length > 0,
    variants: distinct,
    audioTracks: [...audio],
    subtitles: [...subtitles.values()],
  };
}

module.exports = { looksLikeManifest, parse };

File: android/core-extractor/src/main/java/com/obaflix/bridge/HlsManifest.kt
package com.obaflix.bridge

import java.net.URL

/** Uma variante de qualidade declarada em #EXT-X-STREAM-INF. */
data class HlsVariant(
    val bandwidth: Int,
    val resolution: String?,
    val label: String,
)

/** Resumo do que um manifesto HLS oferece ao player. */
data class HlsMediaInfo(
    val isMaster: Boolean,
    val variants: List<HlsVariant> = emptyList(),
    val audioTracks: List<String> = emptyList(),
    val subtitles: List<SubtitleTrack> = emptyList(),
)

/**
 * Leitura mínima de manifestos HLS para comparar servidores antes de escolher um.
 *
 * O extrator antigo aceitava a primeira fonte que respondia, sem saber se ela
 * tinha uma única qualidade ou um master com áudio e legendas. Com este parser a
 * escolha passa a ser feita pelo que o manifesto realmente entrega.
 */
object HlsManifest {

    fun looksLikeManifest(text: String): Boolean = text.trimStart().startsWith("#EXTM3U")

    /**
     * Divide a lista de atributos de uma tag EXT-X preservando vírgulas dentro de
     * valores entre aspas, como em CODECS="avc1.4d401f,mp4a.40.2".
     */
    private fun splitAttributes(raw: String): Map<String, String> {
        val parts = mutableListOf<String>()
        val current = StringBuilder()
        var quoted = false
        for (char in raw) {
            when {
                char == '"' -> { quoted = !quoted; current.append(char) }
                char == ',' && !quoted -> { parts.add(current.toString()); current.setLength(0) }
                else -> current.append(char)
            }
        }
        if (current.isNotEmpty()) parts.add(current.toString())

        val values = linkedMapOf<String, String>()
        for (part in parts) {
            val eq = part.indexOf('=')
            if (eq <= 0) continue
            val key = part.substring(0, eq).trim().uppercase()
            val value = part.substring(eq + 1).trim().trim('"')
            if (key.isNotEmpty()) values[key] = value
        }
        return values
    }

    private fun qualityLabel(resolution: String?, bandwidth: Int): String {
        val height = resolution?.substringAfter('x', "")?.trim()?.toIntOrNull()
        if (height != null && height > 0) return "${height}p"
        if (bandwidth > 0) return "${bandwidth / 1000} kbps"
        return "auto"
    }

    fun parse(text: String, baseUrl: String): HlsMediaInfo {
        if (!looksLikeManifest(text)) return HlsMediaInfo(isMaster = false)

        val variants = mutableListOf<HlsVariant>()
        val audio = linkedSetOf<String>()
        val subtitles = linkedMapOf<String, SubtitleTrack>()

        text.replace("\r\n", "\n").replace('\r', '\n').split("\n").forEach { line ->
            val trimmed = line.trim()
            when {
                trimmed.startsWith("#EXT-X-STREAM-INF:", ignoreCase = true) -> {
                    val attrs = splitAttributes(trimmed.substringAfter(':'))
                    val bandwidth = attrs["BANDWIDTH"]?.toIntOrNull()
                        ?: attrs["AVERAGE-BANDWIDTH"]?.toIntOrNull()
                        ?: 0
                    val resolution = attrs["RESOLUTION"]
                    variants.add(HlsVariant(bandwidth, resolution, qualityLabel(resolution, bandwidth)))
                }

                trimmed.startsWith("#EXT-X-MEDIA:", ignoreCase = true) -> {
                    val attrs = splitAttributes(trimmed.substringAfter(':'))
                    val name = attrs["NAME"]?.takeIf { it.isNotBlank() }
                        ?: attrs["LANGUAGE"]?.takeIf { it.isNotBlank() }
                    when (attrs["TYPE"]?.uppercase()) {
                        "AUDIO" -> name?.let { audio.add(it) }
                        "SUBTITLES", "CLOSED-CAPTIONS" -> {
                            val uri = attrs["URI"]?.takeIf { it.isNotBlank() } ?: return@forEach
                            val resolved = runCatching { URL(URL(baseUrl), uri).toString() }.getOrNull()
                                ?: return@forEach
                            subtitles.putIfAbsent(resolved, SubtitleTrack(resolved, name ?: "Legenda"))
                        }
                    }
                }
            }
        }

        return HlsMediaInfo(
            isMaster = variants.isNotEmpty(),
            variants = variants.distinctBy { it.label }.sortedByDescending { it.bandwidth },
            audioTracks = audio.toList(),
            subtitles = subtitles.values.toList(),
        )
    }
}
